### ISSUE TITLE
Lint idea: top-level `final`/`const` variable defined without explicit type

### ISSUE THREAD

**Issue by GregoryConrad** [dart-lang/linter#5100]
This would be a nice-to-have _opt-in_ lint. Can helpful when handling more complex types.

> GOOD:
> ```dart
> const int i = 0;
> final ComplexType foo = createThing();
> ```
> BAD:
> ```dart
> const i = 0;
> final foo = createThing();
> ```

Rationale: increases code readability (especially for code reviews, when you don't have an IDE handy) by having type information specified explicitly in the source code. While it is sometimes obvious (like in the case of the `const int i = 0` above), it is fairly useful when dealing with non-obvious types.

**Comment by srawlins** [comment]
Thanks for the request. It isn't clear if you are concerned specifically about local variables or top-level variables, but we have the following existing rules:

* [always_specify_types](https://dart.dev/tools/linter-rules/always_specify_types) - I think this achieves what you want.
* [omit_local_variable_types](https://dart.dev/tools/linter-rules/omit_local_variable_types)
* [omit_obvious_local_variable_types](https://dart.dev/tools/linter-rules/omit_obvious_local_variable_types)
* [specify_nonobvious_local_variable_types](https://dart.dev/tools/linter-rules/specify_nonobvious_local_variable_types)
* [type_annotate_public_apis](https://dart.dev/tools/linter-rules/type_annotate_public_apis)

What you are asking for might be akin to the [](https://dart.dev/effective-dart/design#do-type-annotate-fields-and-top-level-variables-if-the-type-isnt-obvious) rule, for which we appear to not have a backing lint rule.
